src/libraries/System.Private.CoreLib/src/System/Threading/ExecutionContext.cs

```diff
@@ -296,24 +270,45 @@ internal static void RunFromThreadPoolDispatchLoop(Thread threadPoolThread, Exec
                 edi = ExceptionDispatchInfo.Capture(ex);
             }
 
-            // Enregister threadPoolThread as it crossed EH, and use enregistered variable
-            Thread currentThread = threadPoolThread;
-
-            ExecutionContext? currentExecutionCtx = currentThread._executionContext;
+            ExecutionContext? currentExecutionCtx = threadPoolThread._executionContext;
 
             // Restore changed SynchronizationContext back to Default
-            currentThread._synchronizationContext = null;
+            threadPoolThread._synchronizationContext = null;
             if (currentExecutionCtx != null)
             {
                 // The EC always needs to be reset for this overload, as it will flow back to the caller if it performs
                 // extra work prior to returning to the Dispatch loop. For example for Task-likes it will flow out of await points
-                RestoreChangedContextToThread(currentThread, contextToRestore: null, currentExecutionCtx);
+                RestoreChangedContextToThread(threadPoolThread, contextToRestore: null, currentExecutionCtx);
             }
 
             // If exception was thrown by callback, rethrow it now original contexts are restored
             edi?.Throw();
         }
 
+        internal static void RunForThreadPoolUnsafe(ExecutionContext executionContext, ContextCallback callback, object state, Thread threadPoolThread)
+        {
+            Debug.Assert(threadPoolThread == Thread.CurrentThread);
+            // We aren't running in try/catch as if an exception is directly thrown on the ThreadPool either process
+            // will crash or its a ThreadAbortException.
```

Comment:
What about `<legacyUnhandledExceptionPolicy enabled="1"/>`?  Or more generally host policy?

Reply:
Ok, then can have ThreadPool catch the exception; do the reset and rethrow; means can simplify a few other paths also.

Reply:
Alas can't do it for other path as Task catches the exception and stashes it as well as doing other things, so need to restore before that.